src/tests/end2end/DepthStencilStateTests.cpp

```diff
@@ -0,0 +1,512 @@
+                    .SetBufferViews(0, 1, &view)
+                    .GetResult();
+
+                nxt::RenderPipeline pipeline = device.CreateRenderPipelineBuilder()
+                    .SetSubpass(renderpass, 0)
+                    .SetLayout(pipelineLayout)
+                    .SetStage(nxt::ShaderStage::Vertex, vsModule, "main")
+                    .SetStage(nxt::ShaderStage::Fragment, fsModule, "main")
+                    .SetDepthStencilState(test.depthStencilState)
+                    .GetResult();
+
+                builder.SetRenderPipeline(pipeline)
+                    .SetStencilReference(test.stencil)
+                    .SetBindGroup(0, bindGroup)
+                    .DrawArrays(6, 1, 0, 0);
+            }
+
+            nxt::CommandBuffer commands = builder
+                .EndRenderSubpass()
+                .EndRenderPass()
+                .GetResult();
+
+            queue.Submit(1, &commands);
+
+            EXPECT_PIXEL_RGBA8_EQ(expectedFront, renderTarget, kRTSize / 4, kRTSize / 2) << "Front face check failed";
+            EXPECT_PIXEL_RGBA8_EQ(expectedBack, renderTarget, 3 * kRTSize / 4, kRTSize / 2) << "Back face check failed";
+        }
+
+        void DoTest(const std::vector<TestSpec> &testParams, const RGBA8& expected) {
+            DoTest(testParams, expected, expected);
+        }
+
+        nxt::RenderPass renderpass;
+        nxt::Texture renderTarget;
+        nxt::Texture depthTexture;
+        nxt::TextureView renderTargetView;
+        nxt::TextureView depthTextureView;
+        nxt::Framebuffer framebuffer;
+        nxt::ShaderModule vsModule;
+        nxt::ShaderModule fsModule;
+        nxt::BindGroupLayout bindGroupLayout;
+        nxt::PipelineLayout pipelineLayout;
+};
+
+// Test compilation and usage of the fixture
+TEST_P(DepthStencilStateTest, Basic) {
+    nxt::DepthStencilState state = device.CreateDepthStencilStateBuilder()
+        .GetResult();
+
+    DoTest({
+        { state, RGBA8(0, 255, 0, 255), 0.5f, 0u },
+    }, RGBA8(0, 255, 0, 255));
+}
+
+// Test defaults: depth and stencil tests disabled
+TEST_P(DepthStencilStateTest, Default) {
+    nxt::DepthStencilState state = device.CreateDepthStencilStateBuilder()
+        .GetResult();
+    
+    TestSpec specs[3] = {
+        { state, RGBA8(255, 0, 0, 255), 0.0f, 0u },
+        { state, RGBA8(0, 255, 0, 255), 0.5f, 0u },
+        { state, RGBA8(0, 0, 255, 255), 1.0f, 0u },
+    };
+
+    // Test that for all combinations, the last triangle drawn is the one visible
+    // We check against three triangles because the stencil test may modify results
+    for (uint32_t last = 0; last < 3; ++last) {
+        uint32_t i = last != 0 ? 0 : 1;
+        uint32_t j = 3 - (i + last);
+        DoTest({ specs[i], specs[j], specs[last] }, specs[last].color);
+        DoTest({ specs[j], specs[i], specs[last] }, specs[last].color);
+    }
+}
+
+TEST_P(DepthStencilStateTest, DepthAlways) {
```

Comment:
note: Ok for not having comment before these tests because they are all the same but can you put one before the block of tests? Same in other places in this file.